From 0ced8a5e3983bd6356598a43bee79914821103d8 Mon Sep 17 00:00:00 2001
From: Robert Lai <c_rlai@codeaurora.org>
Date: Mon, 31 Aug 2015 15:46:48 -0700
Subject: [PATCH] QCamera2: Update mm-qcamera-app command line test app

Add support for Meta stream info
Update buffer dump
Add targets in makefile
Fixed preview, snapshot in menu mode
plus fixed start previw in regression mode

Change-Id: I761aa9c2fa70e18cb953cb5b258bd4c32434653b
---
 QCamera2/stack/mm-camera-test/Android.mk           |  12 +
 QCamera2/stack/mm-camera-test/inc/mm_qcamera_app.h |  56 ++-
 QCamera2/stack/mm-camera-test/src/mm_qcamera_app.c | 422 +++++++++++++++++++--
 .../mm-camera-test/src/mm_qcamera_main_menu.c      |  21 +-
 .../stack/mm-camera-test/src/mm_qcamera_preview.c  |  34 +-
 .../stack/mm-camera-test/src/mm_qcamera_snapshot.c |  21 +-
 .../mm-camera-test/src/mm_qcamera_unit_test.c      |  16 +-
 .../stack/mm-camera-test/src/mm_qcamera_video.c    |  33 +-
 8 files changed, 512 insertions(+), 103 deletions(-)

diff --git a/QCamera2/stack/mm-camera-test/Android.mk b/QCamera2/stack/mm-camera-test/Android.mk
index 1d50614..5e4cf69 100644
--- a/QCamera2/stack/mm-camera-test/Android.mk
+++ b/QCamera2/stack/mm-camera-test/Android.mk
@@ -90,6 +90,18 @@ else ifneq (,$(filter msm8660,$(TARGET_BOARD_PLATFORM)))
         LOCAL_CFLAGS += -DCAMERA_ION_FALLBACK_HEAP_ID=ION_IOMMU_HEAP_ID # EBI
         LOCAL_CFLAGS += -DCAMERA_GRALLOC_CACHING_ID=0
         LOCAL_CFLAGS += -DNUM_RECORDING_BUFFERS=5
+else ifeq ($(TARGET_BOARD_PLATFORM),msm8992)
+        LOCAL_CFLAGS += -DCAMERA_GRALLOC_HEAP_ID=GRALLOC_USAGE_PRIVATE_IOMMU_HEAP
+        LOCAL_CFLAGS += -DCAMERA_GRALLOC_FALLBACK_HEAP_ID=GRALLOC_USAGE_PRIVATE_IOMMU_HEAP
+        LOCAL_CFLAGS += -DCAMERA_ION_FALLBACK_HEAP_ID=ION_IOMMU_HEAP_ID
+        LOCAL_CFLAGS += -DCAMERA_GRALLOC_CACHING_ID=0
+        LOCAL_CFLAGS += -DNUM_RECORDING_BUFFERS=9
+else ifeq ($(TARGET_BOARD_PLATFORM),msm8996)
+        LOCAL_CFLAGS += -DCAMERA_GRALLOC_HEAP_ID=GRALLOC_USAGE_PRIVATE_IOMMU_HEAP
+        LOCAL_CFLAGS += -DCAMERA_GRALLOC_FALLBACK_HEAP_ID=GRALLOC_USAGE_PRIVATE_IOMMU_HEAP
+        LOCAL_CFLAGS += -DCAMERA_ION_FALLBACK_HEAP_ID=ION_IOMMU_HEAP_ID
+        LOCAL_CFLAGS += -DCAMERA_GRALLOC_CACHING_ID=0
+        LOCAL_CFLAGS += -DNUM_RECORDING_BUFFERS=9
 else
         LOCAL_CFLAGS += -DCAMERA_GRALLOC_HEAP_ID=GRALLOC_USAGE_PRIVATE_CAMERA_HEAP
         LOCAL_CFLAGS += -DCAMERA_GRALLOC_FALLBACK_HEAP_ID=GRALLOC_USAGE_PRIVATE_CAMERA_HEAP # Don't Care
diff --git a/QCamera2/stack/mm-camera-test/inc/mm_qcamera_app.h b/QCamera2/stack/mm-camera-test/inc/mm_qcamera_app.h
index f1c688b..4568b48 100644
--- a/QCamera2/stack/mm-camera-test/inc/mm_qcamera_app.h
+++ b/QCamera2/stack/mm-camera-test/inc/mm_qcamera_app.h
@@ -92,6 +92,16 @@
 #define FALSE 0
 #endif
 
+#define QCAMAPP_DUMP_FRM_PREVIEW    1
+#define QCAMAPP_DUMP_FRM_VIDEO      (1<<1)
+#define QCAMAPP_DUMP_FRM_SNAPSHOT   (1<<2)
+#define QCAMAPP_DUMP_FRM_THUMBNAIL  (1<<3)
+#define QCAMAPP_DUMP_FRM_RAW        (1<<4)
+#define QCAMAPP_DUMP_FRM_JPEG       (1<<5)
+
+#define QCAMAPP_DUMP_FRM_MASK_ALL    0x000000ff
+
+
 typedef enum {
     TUNE_CMD_INIT,
     TUNE_CMD_GET_LIST,
@@ -178,6 +188,8 @@ typedef struct {
     uint32_t multipleOf;
     mm_camera_app_buf_t s_bufs[MM_CAMERA_MAX_NUM_FRAMES];
     mm_camera_app_buf_t s_info_buf;
+    uint32_t mDumpFrame;
+    uint32_t mDumpSkipCnt;
 } mm_camera_stream_t;
 
 typedef struct {
@@ -207,9 +219,28 @@ typedef struct {
 } USER_INPUT_DISPLAY_T;
 
 typedef struct {
+    cam_dimension_t dim;
+} mm_camera_lib_snapshot_params;
+
+typedef struct {
+    cam_dimension_t dim;
+} mm_camera_lib_video_params;
+
+typedef struct {
+    USER_INPUT_DISPLAY_T preview_params;
+    mm_camera_lib_snapshot_params snapshot_params;
+    mm_camera_lib_video_params video_params;
+    uint8_t zsl_enabled;
+    uint8_t video_enabled;
+    cam_focus_mode_type af_mode;
+    uint32_t featureMask[CAM_STREAM_TYPE_MAX];
+} mm_camera_lib_params;
+
+typedef struct {
     mm_camera_vtbl_t *cam;
     uint8_t num_channels;
     mm_camera_channel_t channels[MM_CHANNEL_TYPE_MAX];
+    mm_camera_lib_params params;
     mm_jpeg_ops_t jpeg_ops;
     uint32_t jpeg_hdl;
     mm_camera_app_buf_t cap_buf;
@@ -224,17 +255,15 @@ typedef struct {
     struct fb_var_screeninfo vinfo;
     struct mdp_overlay data_overlay;
     uint32_t slice_size;
-    uint32_t buffer_width, buffer_height;
+    int32_t buffer_width, buffer_height;
     uint32_t buffer_size;
     cam_format_t buffer_format;
     uint32_t frame_size;
     uint32_t frame_count;
     int encodeJpeg;
-    int zsl_enabled;
     int8_t focus_supported;
     prev_callback user_preview_cb;
     parm_buffer_t *params_buffer;
-    USER_INPUT_DISPLAY_T preview_resolution;
 
     //Reprocess params&stream
     int8_t enable_reproc;
@@ -263,11 +292,6 @@ typedef struct {
     hal_interface_lib_t hal_lib;
 } mm_camera_app_t;
 
-typedef struct {
-    uint32_t width;
-    uint32_t height;
-} mm_camera_lib_snapshot_params;
-
 typedef enum {
     MM_CAMERA_LIB_NO_ACTION = 0,
     MM_CAMERA_LIB_RAW_CAPTURE,
@@ -318,11 +342,6 @@ typedef enum {
 } mm_camera_lib_commands;
 
 typedef struct {
-    int32_t stream_width, stream_height;
-    cam_focus_mode_type af_mode;
-} mm_camera_lib_params;
-
-typedef struct {
   tuneserver_protocol_t *proto;
   int clientsocket_id;
   prserver_protocol_t *pr_proto;
@@ -333,7 +352,6 @@ typedef struct {
 typedef struct {
     mm_camera_app_t app_ctx;
     mm_camera_test_obj_t test_obj;
-    mm_camera_lib_params current_params;
     int stream_running;
     tuningserver_t tsctrl;
 } mm_camera_lib_ctx;
@@ -348,10 +366,9 @@ typedef struct {
 
 extern int mm_app_unit_test_entry(mm_camera_app_t *cam_app);
 extern int mm_app_dual_test_entry(mm_camera_app_t *cam_app);
-extern void mm_app_dump_frame(mm_camera_buf_def_t *frame,
-                              char *name,
-                              char *ext,
-                              uint32_t frame_idx);
+extern int mm_app_dump_frame(mm_camera_test_obj_t *testObj, mm_camera_stream_t *stream,
+        mm_camera_buf_def_t *frame, uint32_t dump_type);
+
 extern void mm_app_dump_jpeg_frame(const void * data,
                                    size_t size,
                                    char* name,
@@ -476,7 +493,8 @@ int32_t mm_camera_load_tuninglibrary(
   mm_camera_tuning_lib_params_t *tuning_param);
 int mm_camera_lib_set_preview_usercb(
   mm_camera_lib_handle *handle, prev_callback cb);
-//
+int32_t setStreamConfigure(mm_camera_test_obj_t *testObj,
+        int8_t resetConfig, int8_t isCapture);
 
 int mm_app_start_regression_test(int run_tc);
 int mm_app_load_hal(mm_camera_app_t *my_cam_app);
diff --git a/QCamera2/stack/mm-camera-test/src/mm_qcamera_app.c b/QCamera2/stack/mm-camera-test/src/mm_qcamera_app.c
index 877d9ba..fb113f8 100644
--- a/QCamera2/stack/mm-camera-test/src/mm_qcamera_app.c
+++ b/QCamera2/stack/mm-camera-test/src/mm_qcamera_app.c
@@ -233,36 +233,166 @@ int mm_app_cache_ops(mm_camera_app_meminfo_t *mem_info,
     return ret;
 }
 
-void mm_app_dump_frame(mm_camera_buf_def_t *frame,
-                       char *name,
-                       char *ext,
-                       uint32_t frame_idx)
+
+int mm_app_dump_frame(mm_camera_test_obj_t *testObj, mm_camera_stream_t *stream,
+        mm_camera_buf_def_t *frame, uint32_t dump_type)
 {
-    char file_name[FILENAME_MAX];
-    int file_fd;
-    int i;
-    int offset = 0;
-    if ( frame != NULL) {
-        snprintf(file_name, sizeof(file_name),
-                QCAMERA_DUMP_FRM_LOCATION"test/%s_%04d.%s", name, frame_idx, ext);
-        file_fd = open(file_name, O_RDWR | O_CREAT, 0777);
-        if (file_fd < 0) {
-            CDBG_ERROR("%s: cannot open file %s \n", __func__, file_name);
-        } else {
-            for (i = 0; i < frame->planes_buf.num_planes; i++) {
-                CDBG("%s: saving file from address: %p, data offset: %d, "
-                     "length: %d \n", __func__, frame->buffer,
-                    frame->planes_buf.planes[i].data_offset, frame->planes_buf.planes[i].length);
-                write(file_fd,
-                      (uint8_t *)frame->buffer + offset,
-                      frame->planes_buf.planes[i].length);
-                offset += (int)frame->planes_buf.planes[i].length;
+    int rc = MM_CAMERA_OK;
+    char value[PROPERTY_VALUE_MAX];
+    property_get("persist.qcamapp.dumpimg", value, "0");
+    uint32_t enabled = (uint32_t) atoi(value);
+    uint32_t frm_num = 0;
+    uint32_t skip_mode = 0;
+    cam_dimension_t dim;
+    uint32_t i;
+    int32_t j;
+
+    if (NULL == stream) {
+        ALOGE("%s test object is null", __func__);
+        return -1;
+    }
+
+    uint32_t mDumpFrmCnt = stream->mDumpFrame;
+
+    if((enabled & QCAMAPP_DUMP_FRM_MASK_ALL)) {
+        if((enabled & dump_type) && stream && frame) {
+            frm_num = ((enabled & 0xffff0000) >> 16);
+            if(frm_num == 0) {
+                frm_num = 10; //default 10 frames
             }
+            if(frm_num > 256) {
+                frm_num = 256; //256 buffers cycle around
+            }
+            skip_mode = ((enabled & 0x0000ff00) >> 8);
+            if(skip_mode == 0) {
+                skip_mode = 1; //no-skip
+            }
+            if(stream->mDumpSkipCnt == 0)
+                stream->mDumpSkipCnt = 1;
 
-            close(file_fd);
-            CDBG("dump %s", file_name);
+            if( stream->mDumpSkipCnt % skip_mode == 0) {
+                if((frm_num == 256) && (mDumpFrmCnt >= frm_num)) {
+                    // reset frame count if cycling
+                    mDumpFrmCnt = 0;
+                }
+                if (mDumpFrmCnt <= frm_num) {
+                    char filename[128];
+                    char timeBuf[32];
+                    time_t current_time;
+                    struct tm * timeinfo;
+
+                    memset(timeBuf, 0, sizeof(timeBuf));
+
+                    time (&current_time);
+                    timeinfo = localtime (&current_time);
+                    memset(filename, 0, sizeof(filename));
+
+                    cam_frame_len_offset_t offset;
+                    memset(&offset, 0, sizeof(cam_frame_len_offset_t));
+                    offset = stream->offset;
+
+                    memset(&dim, 0, sizeof(dim));
+
+                    if (NULL != timeinfo) {
+                        strftime(timeBuf, sizeof(timeBuf),
+                                QCAMERA_DUMP_FRM_LOCATION "%Y%m%d%H%M%S", timeinfo);
+                    }
+                    switch (dump_type) {
+                    case QCAMAPP_DUMP_FRM_PREVIEW:
+                        {
+                            dim.width = testObj->params.preview_params.user_input_display_width;
+                            dim.height = testObj->params.preview_params.user_input_display_height;
+                            snprintf(filename, sizeof(filename), "%s_%dp_%dx%d_%d.yuv",
+                                     timeBuf, mDumpFrmCnt, dim.width, dim.height, frame->frame_idx);
+                        }
+                        break;
+                    case QCAMAPP_DUMP_FRM_THUMBNAIL:
+                        {
+                            dim.width = testObj->params.preview_params.user_input_display_width;
+                            dim.height = testObj->params.preview_params.user_input_display_height;
+                            snprintf(filename, sizeof(filename), "%s_%dt_%dx%d_%d.yuv",
+                                     timeBuf, mDumpFrmCnt, dim.width, dim.height, frame->frame_idx);
+                        }
+                        break;
+                    case QCAMAPP_DUMP_FRM_SNAPSHOT:
+                        {
+                            dim = testObj->params.snapshot_params.dim;
+                            snprintf(filename, sizeof(filename), "%s_%ds_%dx%d_%d.yuv", timeBuf,
+                                    mDumpFrmCnt,
+                                    dim.width,
+                                    dim.height,
+                                    frame->frame_idx);
+                        }
+                        break;
+                    case QCAMAPP_DUMP_FRM_VIDEO:
+                        {
+                            dim = testObj->params.video_params.dim;
+                            snprintf(filename, sizeof(filename), "%s_%dv_%dx%d_%d.yuv", timeBuf,
+                                     mDumpFrmCnt, dim.width, dim.height, frame->frame_idx);
+                        }
+                        break;
+                    case QCAMAPP_DUMP_FRM_RAW:
+                        {
+                            dim = testObj->params.snapshot_params.dim;
+                            snprintf(filename, sizeof(filename), "%s_%dr_%dx%d_%d.raw",
+                                    timeBuf, mDumpFrmCnt,
+                                    dim.width,
+                                    dim.height,
+                                    frame->frame_idx);
+                        }
+                        break;
+                    case QCAMAPP_DUMP_FRM_JPEG:
+                        {
+                            dim = testObj->params.snapshot_params.dim;
+                            snprintf(filename, sizeof(filename), "%s_%dj_%dx%d_%d.yuv",
+                                    timeBuf, mDumpFrmCnt,
+                                    dim.width,
+                                    dim.height,
+                                    frame->frame_idx);
+                        }
+                        break;
+                    default:
+                        ALOGE("%s: Not supported for dumping stream type %d",
+                              __func__, dump_type);
+                        return -1;
+                    }
+                    CDBG_HIGH("%s Dump to file = %s",__func__,filename);
+                    int file_fd = open(filename, O_RDWR | O_CREAT, 0777);
+                    ssize_t written_len = 0;
+                    if (file_fd > 0) {
+                        void *data = NULL;
+
+                        fchmod(file_fd, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+                        for (i = 0; i < offset.num_planes; i++) {
+                            uint32_t index = offset.mp[i].offset;
+                            if (i > 0) {
+                                index += offset.mp[i-1].len;
+                            }
+                            for (j = 0; j < offset.mp[i].height; j++) {
+                                data = (void *)((uint8_t *)frame->buffer + index);
+                                written_len += write(file_fd, data,
+                                        (size_t)offset.mp[i].width);
+                                index += (uint32_t)offset.mp[i].stride;
+                            }
+                        }
+
+                        CDBG_HIGH("%s: written number of bytes %ld\n",
+                            __func__, written_len);
+                        close(file_fd);
+                    } else {
+                        ALOGE("%s: fail t open file for image dumping", __func__);
+                    }
+                    mDumpFrmCnt++;
+                }
+            }
+            stream->mDumpSkipCnt++;
         }
+    } else {
+        mDumpFrmCnt = 0;
     }
+    stream->mDumpFrame = mDumpFrmCnt;
+
+    return rc;
 }
 
 void mm_app_dump_jpeg_frame(const void * data, size_t size, char* name,
@@ -501,11 +631,13 @@ int mm_app_open(mm_camera_app_t *cam_app,
 
     rc = cam_app->hal_lib.mm_camera_open((uint8_t)cam_id, &(test_obj->cam));
     if(rc) {
-        CDBG_ERROR("%s:dev open error. rc = %d, vtbl = %p\n", __func__, rc, test_obj->cam);
+        CDBG_ERROR("%s:dev open error. rc = %d, vtbl = %p\n",
+                __func__, rc, test_obj->cam);
         return -MM_CAMERA_E_GENERAL;
     }
 
-    CDBG("Open Camera id = %d handle = %d", cam_id, test_obj->cam->camera_handle);
+    CDBG("%s: Open Camera id = %d handle = %d",
+            __func__, cam_id, test_obj->cam->camera_handle);
 
     /* alloc ion mem for capability buf */
     memset(&offset_info, 0, sizeof(offset_info));
@@ -1638,14 +1770,15 @@ ERROR:
  *  Return: >=0 on success, -1 on failure.
  **/
 int tuneserver_capture(mm_camera_lib_handle *lib_handle,
-                       mm_camera_lib_snapshot_params *dim)
+                       mm_camera_lib_snapshot_params *param)
 {
     int rc = 0;
+    cam_dimension_t *dim = &param->dim;
 
-    printf("Take jpeg snapshot\n");
+    printf("%s:Take jpeg snapshot\n", __func__);
     if ( lib_handle->stream_running ) {
 
-        if ( lib_handle->test_obj.zsl_enabled) {
+        if ( lib_handle->test_obj.params.zsl_enabled) {
             if ( NULL != dim) {
                 if ( ( lib_handle->test_obj.buffer_width != dim->width) ||
                      ( lib_handle->test_obj.buffer_height = dim->height ) ) {
@@ -1675,6 +1808,7 @@ int tuneserver_capture(mm_camera_lib_handle *lib_handle,
             mm_camera_app_wait();
         } else {
             // For standard 2D capture streaming has to be disabled first
+
             rc = mm_camera_lib_stop_stream(lib_handle);
             if (rc != MM_CAMERA_OK) {
                 CDBG_ERROR("%s: mm_camera_lib_stop_stream() err=%d\n",
@@ -1686,6 +1820,7 @@ int tuneserver_capture(mm_camera_lib_handle *lib_handle,
                 lib_handle->test_obj.buffer_width = dim->width;
                 lib_handle->test_obj.buffer_height = dim->height;
             }
+
             rc = mm_app_start_capture(&lib_handle->test_obj, 1);
             if (rc != MM_CAMERA_OK) {
                 CDBG_ERROR("%s: mm_app_start_capture() err=%d\n",
@@ -1779,6 +1914,18 @@ int32_t mm_camera_load_tuninglibrary(mm_camera_tuning_lib_params_t *tuning_param
   return 0;
 }
 
+int mm_app_init_param(mm_camera_lib_params *params)
+{
+    int rc = MM_CAMERA_OK;
+
+    params->snapshot_params.dim.width = DEFAULT_SNAPSHOT_WIDTH;
+    params->snapshot_params.dim.height= DEFAULT_SNAPSHOT_HEIGHT;
+    params->preview_params.user_input_display_width = DEFAULT_PREVIEW_WIDTH;
+    params->preview_params.user_input_display_height = DEFAULT_PREVIEW_HEIGHT;
+    params->af_mode = CAM_FOCUS_MODE_AUTO; // Default to auto focus mode
+    return rc;
+}
+
 int mm_camera_lib_open(mm_camera_lib_handle *handle, int cam_id)
 {
     int rc = MM_CAMERA_OK;
@@ -1799,9 +1946,9 @@ int mm_camera_lib_open(mm_camera_lib_handle *handle, int cam_id)
     handle->test_obj.buffer_width = DEFAULT_PREVIEW_WIDTH;
     handle->test_obj.buffer_height = DEFAULT_PREVIEW_HEIGHT;
     handle->test_obj.buffer_format = DEFAULT_SNAPSHOT_FORMAT;
-    handle->current_params.stream_width = DEFAULT_SNAPSHOT_WIDTH;
-    handle->current_params.stream_height = DEFAULT_SNAPSHOT_HEIGHT;
-    handle->current_params.af_mode = CAM_FOCUS_MODE_AUTO; // Default to auto focus mode
+
+    rc = mm_app_init_param(&handle->test_obj.params);
+
     rc = mm_app_open(&handle->app_ctx, (uint8_t)cam_id, &handle->test_obj);
     if (rc != MM_CAMERA_OK) {
         CDBG_ERROR("%s:mm_app_open() cam_idx=%d, err=%d\n",
@@ -1822,6 +1969,191 @@ EXIT:
     return rc;
 }
 
+int32_t updateFeatureMask(mm_camera_test_obj_t *test_obj,
+        cam_stream_type_t stream_type)
+{
+    int32_t rc = MM_CAMERA_OK;
+    uint32_t feature_mask = 0;
+
+    cam_capability_t *caps = (cam_capability_t *)test_obj->cap_buf.mem_info.data;
+
+    if (stream_type >= CAM_STREAM_TYPE_MAX) {
+        ALOGE("%s: Error!! stream type: %d not valid", __func__, stream_type);
+        return -1;
+    }
+
+    if(!((test_obj->params.zsl_enabled)
+            && (stream_type == CAM_STREAM_TYPE_SNAPSHOT))) {
+        CDBG("caps->min_required_pp_mask = %d",caps->min_required_pp_mask);
+        if (caps->min_required_pp_mask & CAM_QCOM_FEATURE_SHARPNESS) {
+            feature_mask |= CAM_QCOM_FEATURE_SHARPNESS;
+        }
+
+        if (caps->min_required_pp_mask & CAM_QCOM_FEATURE_EFFECT) {
+            feature_mask |= CAM_QCOM_FEATURE_EFFECT;
+        }
+    }
+    test_obj->params.featureMask[stream_type] = feature_mask;
+    CDBG("%s: feature_mask = %d", __func__, feature_mask);
+    return rc;
+}
+
+int32_t sendStreamConfigInfo(mm_camera_test_obj_t *test_obj,
+        cam_stream_size_info_t stream_config_info) {
+    int32_t rc = MM_CAMERA_OK;
+    if(initBatchUpdate(test_obj) < 0 ) {
+        ALOGE("%s:Failed to initialize group update table", __func__);
+        return -1;
+    }
+
+    if (ADD_SET_PARAM_ENTRY_TO_BATCH(test_obj->parm_buf.mem_info.data,
+            CAM_INTF_META_STREAM_INFO, stream_config_info)) {
+        CDBG_ERROR("%s:Meta stream info is not added to batch\n", __func__);
+        return -1;
+    }
+
+    rc = commitSetBatch(test_obj);
+    if (rc != MM_CAMERA_OK) {
+        ALOGE("%s:Failed to set stream info parm", __func__);
+        return rc;
+    }
+    return rc;
+}
+
+
+int32_t setStreamConfigure(mm_camera_test_obj_t *testObj,
+        int8_t resetConfig, int8_t isCapture) {
+
+    int32_t rc = MM_CAMERA_OK;
+    uint32_t k = 0;
+    cam_stream_size_info_t stream_config_info;
+    int width,height;
+
+    CDBG("%s : E", __func__);
+    memset(&stream_config_info, 0, sizeof(stream_config_info));
+    stream_config_info.num_streams = 0;
+
+    if (resetConfig) {
+        CDBG_HIGH("%s: Reset stream config!!", __func__);
+        rc = sendStreamConfigInfo(testObj,stream_config_info);
+        return rc;
+    }
+
+    if (testObj->params.zsl_enabled) {
+        stream_config_info.type[stream_config_info.num_streams] =
+            CAM_STREAM_TYPE_PREVIEW;
+
+        if ((testObj->params.preview_params.user_input_display_width == 0) ||
+            ( testObj->params.preview_params.user_input_display_height == 0)) {
+            width = DEFAULT_PREVIEW_WIDTH;
+            height = DEFAULT_PREVIEW_HEIGHT;
+        } else {
+            width = testObj->params.preview_params.user_input_display_width;
+            height = testObj->params.preview_params.user_input_display_height;
+        }
+        stream_config_info.stream_sizes[stream_config_info.num_streams].width =
+            width;
+        stream_config_info.stream_sizes[stream_config_info.num_streams].height =
+            height;
+        updateFeatureMask(testObj,CAM_STREAM_TYPE_PREVIEW);
+        stream_config_info.postprocess_mask[stream_config_info.num_streams] =
+                testObj->params.featureMask[CAM_STREAM_TYPE_PREVIEW];
+        stream_config_info.num_streams++;
+
+        stream_config_info.type[stream_config_info.num_streams] =
+            CAM_STREAM_TYPE_SNAPSHOT;
+
+        width = testObj->params.snapshot_params.dim.width;
+        height = testObj->params.snapshot_params.dim.height;
+
+        stream_config_info.stream_sizes[stream_config_info.num_streams].width =
+            width;
+        stream_config_info.stream_sizes[stream_config_info.num_streams].height =
+            height;
+        updateFeatureMask(testObj,CAM_STREAM_TYPE_SNAPSHOT);
+        stream_config_info.postprocess_mask[stream_config_info.num_streams] =
+                testObj->params.featureMask[CAM_STREAM_TYPE_SNAPSHOT];
+        stream_config_info.num_streams++;
+    }  else if (!isCapture){
+        if (testObj->params.video_enabled) {
+            stream_config_info.type[stream_config_info.num_streams] =
+                    CAM_STREAM_TYPE_SNAPSHOT;
+            width = testObj->params.snapshot_params.dim.width;
+            height = testObj->params.snapshot_params.dim.height;
+
+            stream_config_info.stream_sizes[stream_config_info.num_streams].width
+                = width;
+            stream_config_info.stream_sizes[stream_config_info.num_streams].height
+                = height;
+            updateFeatureMask(testObj,CAM_STREAM_TYPE_SNAPSHOT);
+            stream_config_info.postprocess_mask[stream_config_info.num_streams]
+                = testObj->params.featureMask[CAM_STREAM_TYPE_SNAPSHOT];
+            stream_config_info.num_streams++;
+
+            stream_config_info.type[stream_config_info.num_streams] =
+                    CAM_STREAM_TYPE_VIDEO;
+            width = testObj->params.video_params.dim.width;
+            height = testObj->params.video_params.dim.height;
+
+            stream_config_info.stream_sizes[stream_config_info.num_streams].width
+                = width;
+            stream_config_info.stream_sizes[stream_config_info.num_streams].height
+                = height;
+            updateFeatureMask(testObj,CAM_STREAM_TYPE_VIDEO);
+            stream_config_info.postprocess_mask[stream_config_info.num_streams]
+                = testObj->params.featureMask[CAM_STREAM_TYPE_VIDEO];
+            stream_config_info.num_streams++;
+        }
+
+        stream_config_info.type[stream_config_info.num_streams] =
+                CAM_STREAM_TYPE_PREVIEW;
+
+        if ((testObj->params.preview_params.user_input_display_width == 0) ||
+            ( testObj->params.preview_params.user_input_display_height == 0)) {
+            width = DEFAULT_PREVIEW_WIDTH;
+            height = DEFAULT_PREVIEW_HEIGHT;
+        } else {
+            width = testObj->params.preview_params.user_input_display_width;
+            height = testObj->params.preview_params.user_input_display_height;
+        }
+        stream_config_info.stream_sizes[stream_config_info.num_streams].width =
+            width;
+        stream_config_info.stream_sizes[stream_config_info.num_streams].height =
+             height;
+        updateFeatureMask(testObj,CAM_STREAM_TYPE_PREVIEW);
+        stream_config_info.postprocess_mask[stream_config_info.num_streams] =
+            testObj->params.featureMask[CAM_STREAM_TYPE_PREVIEW];
+        stream_config_info.num_streams++;
+
+    } else {
+        stream_config_info.type[stream_config_info.num_streams] =
+                    CAM_STREAM_TYPE_SNAPSHOT;
+        width = testObj->params.snapshot_params.dim.width;
+        height = testObj->params.snapshot_params.dim.height;
+
+        stream_config_info.stream_sizes[stream_config_info.num_streams].width =
+            width;
+        stream_config_info.stream_sizes[stream_config_info.num_streams].height =
+            height;
+        updateFeatureMask(testObj,CAM_STREAM_TYPE_SNAPSHOT);
+        stream_config_info.postprocess_mask[stream_config_info.num_streams] =
+            testObj->params.featureMask[CAM_STREAM_TYPE_SNAPSHOT];
+        stream_config_info.num_streams++;
+    }
+    for (k = 0; k < stream_config_info.num_streams; k++) {
+        CDBG("%s: stream type %d, w x h: %d x %d, pp_mask: 0x%x", __func__,
+                stream_config_info.type[k],
+                stream_config_info.stream_sizes[k].width,
+                stream_config_info.stream_sizes[k].height,
+                stream_config_info.postprocess_mask[k]);
+    }
+
+    rc = sendStreamConfigInfo(testObj,stream_config_info);
+    CDBG("%s : X", __func__);
+    return rc;
+}
+
+
 int mm_camera_lib_start_stream(mm_camera_lib_handle *handle)
 {
     int rc = MM_CAMERA_OK;
@@ -1833,7 +2165,13 @@ int mm_camera_lib_start_stream(mm_camera_lib_handle *handle)
         goto EXIT;
     }
 
-    if ( handle->test_obj.zsl_enabled ) {
+    rc = setStreamConfigure(&handle->test_obj, FALSE, FALSE);
+    if (rc != MM_CAMERA_OK) {
+        CDBG_ERROR("%s:%d setStreamConfigure() failed\n",
+                   __func__, __LINE__);
+        goto EXIT;
+    }
+    if ( handle->test_obj.params.zsl_enabled ) {
         rc = mm_app_start_preview_zsl(&handle->test_obj);
         if (rc != MM_CAMERA_OK) {
             CDBG_ERROR("%s: mm_app_start_preview_zsl() err=%d\n",
@@ -1860,11 +2198,11 @@ int mm_camera_lib_start_stream(mm_camera_lib_handle *handle)
       camera_cap.supported_focus_modes[0] == CAM_FOCUS_MODE_FIXED) {
       CDBG("focus not supported");
       handle->test_obj.focus_supported = 0;
-      handle->current_params.af_mode = CAM_FOCUS_MODE_FIXED;
+      handle->test_obj.params.af_mode = CAM_FOCUS_MODE_FIXED;
     } else {
       handle->test_obj.focus_supported = 1;
     }
-    rc = setFocusMode(&handle->test_obj, handle->current_params.af_mode);
+    rc = setFocusMode(&handle->test_obj, handle->test_obj.params.af_mode);
     if (rc != MM_CAMERA_OK) {
       CDBG_ERROR("%s:autofocus error\n", __func__);
       goto EXIT;
@@ -1885,7 +2223,7 @@ int mm_camera_lib_stop_stream(mm_camera_lib_handle *handle)
         goto EXIT;
     }
 
-    if ( handle->test_obj.zsl_enabled ) {
+    if ( handle->test_obj.params.zsl_enabled ) {
         rc = mm_app_stop_preview_zsl(&handle->test_obj);
         if (rc != MM_CAMERA_OK) {
             CDBG_ERROR("%s: mm_app_stop_preview_zsl() err=%d\n",
@@ -1900,7 +2238,7 @@ int mm_camera_lib_stop_stream(mm_camera_lib_handle *handle)
             goto EXIT;
         }
     }
-
+    setStreamConfigure(&handle->test_obj, TRUE, FALSE);
     handle->stream_running = 0;
 
 EXIT:
@@ -2107,7 +2445,7 @@ int mm_camera_lib_send_command(mm_camera_lib_handle *handle,
         case MM_CAMERA_LIB_ZSL_ENABLE:
             if ( NULL != in_data) {
                 int enable_zsl = *(( int * )in_data);
-                if ( ( enable_zsl != handle->test_obj.zsl_enabled ) &&
+                if ( ( enable_zsl != handle->test_obj.params.zsl_enabled ) &&
                         handle->stream_running ) {
                     rc = mm_camera_lib_stop_stream(handle);
                     if (rc != MM_CAMERA_OK) {
@@ -2115,7 +2453,7 @@ int mm_camera_lib_send_command(mm_camera_lib_handle *handle,
                                    __func__, rc);
                         goto EXIT;
                     }
-                    handle->test_obj.zsl_enabled = enable_zsl;
+                    handle->test_obj.params.zsl_enabled = enable_zsl;
                     rc = mm_camera_lib_start_stream(handle);
                     if (rc != MM_CAMERA_OK) {
                         CDBG_ERROR("%s: mm_camera_lib_start_stream() err=%d\n",
@@ -2123,7 +2461,7 @@ int mm_camera_lib_send_command(mm_camera_lib_handle *handle,
                         goto EXIT;
                     }
                 } else {
-                    handle->test_obj.zsl_enabled = enable_zsl;
+                    handle->test_obj.params.zsl_enabled = enable_zsl;
                 }
             }
             break;
@@ -2212,7 +2550,7 @@ int mm_camera_lib_send_command(mm_camera_lib_handle *handle,
 
         case MM_CAMERA_LIB_SET_FOCUS_MODE: {
             cam_focus_mode_type mode = *((cam_focus_mode_type *)in_data);
-            handle->current_params.af_mode = mode;
+            handle->test_obj.params.af_mode = mode;
             rc = setFocusMode(&handle->test_obj, mode);
             if (rc != MM_CAMERA_OK) {
               CDBG_ERROR("%s:autofocus error\n", __func__);
diff --git a/QCamera2/stack/mm-camera-test/src/mm_qcamera_main_menu.c b/QCamera2/stack/mm-camera-test/src/mm_qcamera_main_menu.c
index d4d785d..a85c6f4 100644
--- a/QCamera2/stack/mm-camera-test/src/mm_qcamera_main_menu.c
+++ b/QCamera2/stack/mm-camera-test/src/mm_qcamera_main_menu.c
@@ -1,4 +1,4 @@
-/* Copyright (c) 2013-2014, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2013-2015, The Linux Foundation. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are
@@ -1675,14 +1675,11 @@ static int submain()
         (int)(sizeof(dimension_tbl)/sizeof(dimension_tbl[0]));
     int i,c;
     mm_camera_lib_snapshot_params snap_dim;
-    snap_dim.width = DEFAULT_SNAPSHOT_WIDTH;
-    snap_dim.height = DEFAULT_SNAPSHOT_HEIGHT;
+    snap_dim.dim.width = DEFAULT_SNAPSHOT_WIDTH;
+    snap_dim.dim.height = DEFAULT_SNAPSHOT_HEIGHT;
     cam_scene_mode_type default_scene= CAM_SCENE_MODE_OFF;
     int set_tintless= 0;
 
-    mm_camera_test_obj_t test_obj;
-    memset(&test_obj, 0, sizeof(mm_camera_test_obj_t));
-
     rc = mm_camera_lib_open(&lib_handle, 0);
     if (rc != MM_CAMERA_OK) {
         CDBG_ERROR("%s:mm_camera_lib_open() err=%d\n", __func__, rc);
@@ -1708,8 +1705,8 @@ static int submain()
             CDBG_ERROR("%s:filter_resolutions()\n", __func__);
             goto ERROR;
         }
-        snap_dim.width = dimension_tbl[i].width;
-        snap_dim.height = dimension_tbl[i].height;
+        snap_dim.dim.width = dimension_tbl[i].width;
+        snap_dim.dim.height = dimension_tbl[i].height;
 
         rc = enableAFR(&lib_handle);
         if (rc != MM_CAMERA_OK) {
@@ -1900,8 +1897,8 @@ static int submain()
                     CDBG_ERROR("%s:filter_resolutions()\n", __func__);
                     goto ERROR;
                 }
-                snap_dim.width = dimension_tbl[i].width;
-                snap_dim.height = dimension_tbl[i].height;
+                snap_dim.dim.width = dimension_tbl[i].width;
+                snap_dim.dim.height = dimension_tbl[i].height;
 
                 rc = enableAFR(&lib_handle);
                 if (rc != MM_CAMERA_OK) {
@@ -1986,8 +1983,8 @@ static int submain()
                 printf("\n Switch snapshot resolution to %dx%d\n",
                        dimension_tbl[action_param].width,
                        dimension_tbl[action_param].height);
-                snap_dim.width = dimension_tbl[action_param].width;
-                snap_dim.height = dimension_tbl[action_param].height;
+                snap_dim.dim.width = dimension_tbl[action_param].width;
+                snap_dim.dim.height = dimension_tbl[action_param].height;
                 break;
 
       case ACTION_START_RECORDING:
diff --git a/QCamera2/stack/mm-camera-test/src/mm_qcamera_preview.c b/QCamera2/stack/mm-camera-test/src/mm_qcamera_preview.c
index 4f6cf86..76038e6 100644
--- a/QCamera2/stack/mm-camera-test/src/mm_qcamera_preview.c
+++ b/QCamera2/stack/mm-camera-test/src/mm_qcamera_preview.c
@@ -129,6 +129,8 @@ static void mm_app_preview_notify_cb(mm_camera_super_buf_t *bufs,
         return;
     }
 
+    CDBG("%s: BEGIN\n", __func__);
+
     frame = bufs->bufs[0];
 
     /* find channel */
@@ -142,6 +144,7 @@ static void mm_app_preview_notify_cb(mm_camera_super_buf_t *bufs,
         CDBG_ERROR("%s: Channel object is NULL ", __func__);
         return;
     }
+
     /* find preview stream */
     for (i = 0; i < channel->num_streams; i++) {
         if (channel->streams[i].s_config.stream_info->stream_type == CAM_STREAM_TYPE_PREVIEW) {
@@ -166,13 +169,9 @@ static void mm_app_preview_notify_cb(mm_camera_super_buf_t *bufs,
     if ( 0 < pme->fb_fd ) {
         mm_app_overlay_display(pme, frame->fd);
     }
-#ifdef DUMP_PRV_IN_FILE
-    {
-        char file_name[64];
-        snprintf(file_name, sizeof(file_name), "P_C%d", pme->cam->camera_handle);
-        mm_app_dump_frame(frame, file_name, "yuv", frame->frame_idx);
-    }
-#endif
+
+    mm_app_dump_frame(pme,p_stream,frame,QCAMAPP_DUMP_FRM_PREVIEW);
+
     if (pme->user_preview_cb) {
         CDBG_ERROR("[DBG] %s, user defined own preview cb. calling it...", __func__);
         pme->user_preview_cb(frame);
@@ -308,10 +307,7 @@ static void mm_app_zsl_notify_cb(mm_camera_super_buf_t *bufs,
 
     if ( 0 < pme->fb_fd ) {
         mm_app_overlay_display(pme, p_frame->fd);
-    }/* else {
-        mm_app_dump_frame(p_frame, "zsl_preview", "yuv", p_frame->frame_idx);
-        mm_app_dump_frame(m_frame, "zsl_main", "yuv", m_frame->frame_idx);
-    }*/
+    }
 
     if ( pme->enable_reproc && ( NULL != pme->reproc_stream ) ) {
 
@@ -459,13 +455,13 @@ mm_camera_stream_t * mm_app_add_preview_stream(mm_camera_test_obj_t *test_obj,
     stream->s_config.stream_info->streaming_mode = CAM_STREAMING_MODE_CONTINUOUS;
     stream->s_config.stream_info->fmt = DEFAULT_PREVIEW_FORMAT;
 
-    if ((test_obj->preview_resolution.user_input_display_width == 0) ||
-           ( test_obj->preview_resolution.user_input_display_height == 0)) {
+    if ((test_obj->params.preview_params.user_input_display_width == 0) ||
+           ( test_obj->params.preview_params.user_input_display_height == 0)) {
         stream->s_config.stream_info->dim.width = DEFAULT_PREVIEW_WIDTH;
         stream->s_config.stream_info->dim.height = DEFAULT_PREVIEW_HEIGHT;
     } else {
-        stream->s_config.stream_info->dim.width = test_obj->preview_resolution.user_input_display_width;
-        stream->s_config.stream_info->dim.height = test_obj->preview_resolution.user_input_display_height;
+        stream->s_config.stream_info->dim.width = test_obj->params.preview_params.user_input_display_width;
+        stream->s_config.stream_info->dim.height = test_obj->params.preview_params.user_input_display_height;
     }
 
     stream->s_config.padding_info = cam_cap->padding_info;
@@ -569,13 +565,7 @@ mm_camera_stream_t * mm_app_add_snapshot_stream(mm_camera_test_obj_t *test_obj,
         stream->s_config.stream_info->num_of_burst = num_burst;
     }
     stream->s_config.stream_info->fmt = DEFAULT_SNAPSHOT_FORMAT;
-    if ( test_obj->buffer_width == 0 || test_obj->buffer_height == 0 ) {
-        stream->s_config.stream_info->dim.width = DEFAULT_SNAPSHOT_WIDTH;
-        stream->s_config.stream_info->dim.height = DEFAULT_SNAPSHOT_HEIGHT;
-    } else {
-        stream->s_config.stream_info->dim.width = (int32_t)test_obj->buffer_width;
-        stream->s_config.stream_info->dim.height = (int32_t)test_obj->buffer_height;
-    }
+    stream->s_config.stream_info->dim = test_obj->params.snapshot_params.dim;
     stream->s_config.padding_info = cam_cap->padding_info;
 
     rc = mm_app_config_stream(test_obj, channel, stream, &stream->s_config);
diff --git a/QCamera2/stack/mm-camera-test/src/mm_qcamera_snapshot.c b/QCamera2/stack/mm-camera-test/src/mm_qcamera_snapshot.c
index b644e03..f1386c5 100644
--- a/QCamera2/stack/mm-camera-test/src/mm_qcamera_snapshot.c
+++ b/QCamera2/stack/mm-camera-test/src/mm_qcamera_snapshot.c
@@ -312,7 +312,7 @@ static void mm_app_snapshot_notify_cb_raw(mm_camera_super_buf_t *bufs,
         goto EXIT;
     }
 
-    mm_app_dump_frame(m_frame, "main", "raw", m_frame->frame_idx);
+    mm_app_dump_frame(pme, m_stream, m_frame, QCAMAPP_DUMP_FRM_RAW);
 
 EXIT:
     for (i=0; i<bufs->num_bufs; i++) {
@@ -382,7 +382,7 @@ static void mm_app_snapshot_notify_cb(mm_camera_super_buf_t *bufs,
         goto error;
     }
 
-    mm_app_dump_frame(m_frame, "main", "yuv", m_frame->frame_idx);
+    mm_app_dump_frame(pme, m_stream, m_frame, QCAMAPP_DUMP_FRM_SNAPSHOT);
 
     /* find postview stream */
     for (i = 0; i < channel->num_streams; i++) {
@@ -400,7 +400,7 @@ static void mm_app_snapshot_notify_cb(mm_camera_super_buf_t *bufs,
             }
         }
         if (NULL != p_frame) {
-            mm_app_dump_frame(p_frame, "postview", "yuv", p_frame->frame_idx);
+            mm_app_dump_frame(pme, p_stream, p_frame, QCAMAPP_DUMP_FRM_THUMBNAIL);
         }
     }
 
@@ -605,6 +605,14 @@ int mm_app_start_capture(mm_camera_test_obj_t *test_obj,
     memset(&attr, 0, sizeof(mm_camera_channel_attr_t));
     attr.notify_mode = MM_CAMERA_SUPER_BUF_NOTIFY_CONTINUOUS;
     attr.max_unmatched_frames = 3;
+
+    rc = setStreamConfigure(test_obj, FALSE, TRUE);
+    if (rc != MM_CAMERA_OK) {
+        CDBG_ERROR("%s: setStreamConfigure() failed\n",
+                   __func__);
+        return rc;
+    }
+
     channel = mm_app_add_channel(test_obj,
                                  MM_CHANNEL_TYPE_CAPTURE,
                                  &attr,
@@ -661,6 +669,13 @@ int mm_app_stop_capture(mm_camera_test_obj_t *test_obj)
         CDBG_ERROR("%s:stop capture channel failed rc=%d\n", __func__, rc);
     }
 
+    rc = setStreamConfigure(test_obj, TRUE, FALSE);
+    if (rc != MM_CAMERA_OK) {
+        CDBG_ERROR("%s:%d: setStreamConfigure() failed\n",
+                   __func__, __LINE__);
+        return rc;
+    }
+
     return rc;
 }
 
diff --git a/QCamera2/stack/mm-camera-test/src/mm_qcamera_unit_test.c b/QCamera2/stack/mm-camera-test/src/mm_qcamera_unit_test.c
index 51f4f5d..b0d7763 100644
--- a/QCamera2/stack/mm-camera-test/src/mm_qcamera_unit_test.c
+++ b/QCamera2/stack/mm-camera-test/src/mm_qcamera_unit_test.c
@@ -1,4 +1,4 @@
-/* Copyright (c) 2013, The Linux Foundation. All rights reserved.
+/* Copyright (c) 2013-2015, The Linux Foundation. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are
@@ -44,6 +44,7 @@ int mm_app_tc_open_close(mm_camera_app_t *cam_app)
     mm_camera_test_obj_t test_obj;
 
     printf("\n Verifying open/close cameras...\n");
+    CDBG("%s: Verifying open/close cameras...", __func__);
     for (i = 0; i < cam_app->num_cameras; i++) {
         memset(&test_obj, 0, sizeof(mm_camera_test_obj_t));
         rc = mm_app_open(cam_app, i, &test_obj);
@@ -62,8 +63,10 @@ int mm_app_tc_open_close(mm_camera_app_t *cam_app)
     }
     if (rc == MM_CAMERA_OK) {
         printf("\nPassed\n");
+        CDBG("%s:Passed", __func__);
     } else {
         printf("\nFailed\n");
+        CDBG_ERROR("%s:Failed", __func__);
     }
     CDBG("%s:END, rc = %d\n", __func__, rc);
     return rc;
@@ -76,6 +79,7 @@ int mm_app_tc_start_stop_preview(mm_camera_app_t *cam_app)
     mm_camera_test_obj_t test_obj;
 
     printf("\n Verifying start/stop preview...\n");
+    CDBG("%s: Verifying start/stop preview...", __func__);
     for (i = 0; i < cam_app->num_cameras; i++) {
         memset(&test_obj, 0, sizeof(mm_camera_test_obj_t));
         rc = mm_app_open(cam_app, i, &test_obj);
@@ -86,6 +90,12 @@ int mm_app_tc_start_stop_preview(mm_camera_app_t *cam_app)
         }
 
         for (j = 0; j < MM_QCAMERA_APP_UTEST_INNER_LOOP; j++) {
+            setStreamConfigure(&test_obj,FALSE,FALSE);
+            if (rc != MM_CAMERA_OK) {
+                CDBG_ERROR("%s: setStreamConfigure() cam_idx=%d, err=%d\n",
+                           __func__, i, rc);
+                break;
+            }
             rc = mm_app_start_preview(&test_obj);
             if (rc != MM_CAMERA_OK) {
                 CDBG_ERROR("%s: mm_app_start_preview() cam_idx=%d, err=%d\n",
@@ -110,8 +120,10 @@ int mm_app_tc_start_stop_preview(mm_camera_app_t *cam_app)
     }
     if (rc == MM_CAMERA_OK) {
         printf("\nPassed\n");
+        CDBG("%s:Passed", __func__);
     } else {
         printf("\nFailed\n");
+        CDBG_ERROR("%s:Failed", __func__);
     }
     CDBG("%s:END, rc = %d\n", __func__, rc);
     return rc;
@@ -685,7 +697,7 @@ int mm_app_unit_test_entry(mm_camera_app_t *cam_app)
         }
     }
 end:
-    printf("nTOTAL_TSET_CASE = %d, NUM_TEST_RAN = %d, rc=%d\n", tc, i, rc);
+    printf("\nTOTAL_TSET_CASE = %d, NUM_TEST_RAN = %d, rc=%d\n", tc, i, rc);
     return rc;
 }
 
diff --git a/QCamera2/stack/mm-camera-test/src/mm_qcamera_video.c b/QCamera2/stack/mm-camera-test/src/mm_qcamera_video.c
index 338105e..31c1d6a 100644
--- a/QCamera2/stack/mm-camera-test/src/mm_qcamera_video.c
+++ b/QCamera2/stack/mm-camera-test/src/mm_qcamera_video.c
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2012-2014, The Linux Foundation. All rights reserved.
+Copyright (c) 2012-2015, The Linux Foundation. All rights reserved.
 
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions are
@@ -34,13 +34,40 @@ static void mm_app_video_notify_cb(mm_camera_super_buf_t *bufs,
                                    void *user_data)
 {
     char file_name[64];
+    int i;
+    mm_camera_channel_t *channel = NULL;
+    mm_camera_stream_t *p_stream = NULL;
     mm_camera_buf_def_t *frame = bufs->bufs[0];
     mm_camera_test_obj_t *pme = (mm_camera_test_obj_t *)user_data;
 
     CDBG("%s: BEGIN - length=%zu, frame idx = %d\n",
          __func__, frame->frame_len, frame->frame_idx);
-    snprintf(file_name, sizeof(file_name), "V_C%d", pme->cam->camera_handle);
-    mm_app_dump_frame(frame, file_name, "yuv", frame->frame_idx);
+
+    /* find channel */
+    for (i = 0; i < MM_CHANNEL_TYPE_MAX; i++) {
+        if (pme->channels[i].ch_id == bufs->ch_id) {
+            channel = &pme->channels[i];
+            break;
+        }
+    }
+    if (NULL == channel) {
+        CDBG_ERROR("%s: Channel object is NULL ", __func__);
+        return;
+    }
+    /* find preview stream */
+    for (i = 0; i < channel->num_streams; i++) {
+        if (channel->streams[i].s_config.stream_info->stream_type == CAM_STREAM_TYPE_VIDEO) {
+            p_stream = &channel->streams[i];
+            break;
+        }
+    }
+
+    if (NULL == p_stream) {
+        CDBG_ERROR("%s: cannot find preview stream", __func__);
+        return;
+    }
+
+    mm_app_dump_frame(pme,p_stream,frame,QCAMAPP_DUMP_FRM_VIDEO);
 
     if (MM_CAMERA_OK != pme->cam->ops->qbuf(bufs->camera_handle,
                                             bufs->ch_id,
-- 
1.9.1

